tools/benchmarksAnalyzer/src/main/kotlin/org/jetbrains/renders/HTMLRender.kt

```diff
@@ -0,0 +1,674 @@
+        builder.append("$indent<$name${renderAttributes()}>\n")
+        for (c in children) {
+            c.render(builder, indent + "  ")
+        }
+        builder.append("$indent</$name>\n")
+    }
+
+    private fun renderAttributes(): String {
+        val builder = StringBuilder()
+        for ((attr, value) in attributes) {
+            builder.append(" $attr=\"$value\"")
+        }
+        return builder.toString()
+    }
+
+    override fun toString(): String {
+        val builder = StringBuilder()
+        render(builder, "")
+        return builder.toString()
+    }
+}
+
+abstract class TagWithText(name: String) : Tag(name) {
+    operator fun String.unaryPlus() {
+        children.add(TextElement(this))
+    }
+}
+
+class HTML : TagWithText("html") {
+    fun head(init: Head.() -> Unit) = initTag(Head(), init)
+    fun body(init: Body.() -> Unit) = initTag(Body(), init)
+}
+
+class Head : TagWithText("head") {
+    fun title(init: Title.() -> Unit) = initTag(Title(), init)
+    fun link(init: Link.() -> Unit) = initTag(Link(), init)
+    fun script(init: Script.() -> Unit) = initTag(Script(), init)
+}
+
+class Title : TagWithText("title")
+class Link : TagWithText("link")
+class Script : TagWithText("script")
+
+abstract class BodyTag(name: String) : TagWithText(name) {
+    fun b(init: B.() -> Unit) = initTag(B(), init)
+    fun p(init: P.() -> Unit) = initTag(P(), init)
+    fun h1(init: H1.() -> Unit) = initTag(H1(), init)
+    fun h2(init: H2.() -> Unit) = initTag(H2(), init)
+    fun h4(init: H4.() -> Unit) = initTag(H4(), init)
+    fun hr(init: HR.() -> Unit) = initTag(HR(), init)
+    fun a(href: String, init: A.() -> Unit) {
+        val a = initTag(A(), init)
+        a.href = href
+    }
+
+    fun img(src: String, init: Image.() -> Unit) {
+        val element = initTag(Image(), init)
+        element.src = src
+    }
+
+    fun table(init: Table.() -> Unit) = initTag(Table(), init)
+    fun div(classAttr: String, init: Div.() -> Unit) = initTag(Div(classAttr), init)
+    fun button(classAttr: String, init: Button.() -> Unit) = initTag(Button(classAttr), init)
+    fun header(classAttr: String, init: Header.() -> Unit) = initTag(Header(classAttr), init)
+    fun span(classAttr: String, init: Span.() -> Unit) = initTag(Span(classAttr), init)
+}
+
+abstract class BodyTagWithClass(name: String, val classAttr: String) : BodyTag(name) {
+    init {
+        attributes["class"] = classAttr
+    }
+}
+
+class Body : BodyTag("body")
+class B : BodyTag("b")
+class P : BodyTag("p")
+class H1 : BodyTag("h1")
+class H2 : BodyTag("h2")
+class H4 : BodyTag("h4")
+class HR : BodyTag("hr")
+class Div(classAttr: String) : BodyTagWithClass("div", classAttr)
+class Header(classAttr: String) : BodyTagWithClass("header", classAttr)
+class Button(classAttr: String) : BodyTagWithClass("button", classAttr)
+class Span(classAttr: String) : BodyTagWithClass("span", classAttr)
+
+class A : BodyTag("a") {
+    var href: String
```

Comment:
It looks like [property delegation](https://kotlinlang.org/docs/reference/delegated-properties.html#storing-properties-in-a-map) is very suitable here:
```
class A : BodyTag("a") {
    var href: String by attributes
}
```